torch/testing/_asserts.py

```diff
@@ -102,20 +102,87 @@ def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Excepti
+    """
+
+    @functools.wraps(check_tensors)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_sparse:
+            return check_tensors(actual, expected, **kwargs)
+
+        if actual._nnz() != expected._nnz():
+            return AssertionError(
+                f"The number of specified values does not match: {actual._nnz()} != {expected._nnz()}"
+            )
+
+        kwargs_equal = dict(kwargs, rtol=0, atol=0)
+        exc = check_tensors(actual._indices(), expected._indices(), **kwargs_equal)
+        if exc:
+            return _amend_error_message(exc, "{}\n\nThe failure occurred for the indices.")
+
+        exc = check_tensors(actual._values(), expected._values(), **kwargs)
+        if exc:
+            return _amend_error_message(exc, "{}\n\nThe failure occurred for the values.")
 
-    All checks are temporary and will be relaxed in the future.
+        return None
+
+    return wrapper
+
+
+def _check_sparse_csr_members_individually(
+    check_tensors: Callable[..., Optional[Exception]]
+) -> Callable[..., Optional[Exception]]:
+    """Decorates strided tensor check functions to individually handle sparse CSR members.
+
+    If the inputs are not sparse CSR, this decorator is a no-op.
+
+    Args:
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Tensor check function for strided
+        tensors.
+    """
+
+    @functools.wraps(check_tensors)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_sparse_csr:
+            return check_tensors(actual, expected, **kwargs)
+
+        kwargs_equal = dict(kwargs, rtol=0, atol=0)
+        exc = check_tensors(actual.crow_indices(), expected.crow_indices(), **kwargs_equal)
+        if exc:
+            return _amend_error_message(exc, "{}\n\nThe failure occurred for the crow_indices.")
+
+        exc = check_tensors(actual.col_indices(), expected.col_indices(), **kwargs_equal)
+        if exc:
+            return _amend_error_message(exc, "{}\n\nThe failure occurred for the col_indices.")
+
+        exc = check_tensors(actual.values(), expected.values(), **kwargs)
+        if exc:
+            return _amend_error_message(exc, "{}\n\nThe failure occurred for the values.")
+
+        return None
+
+    return wrapper
+
+
+def _check_supported_tensor(input: Tensor) -> Optional[UsageError]:  # type: ignore[valid-type]
+    """Checks if the tensor is supported by the current infrastructure.
 
     Returns:
         (Optional[UsageError]): If check did not pass.
     """
     if input.is_quantized:
         return UsageError("Comparison for quantized tensors is not supported yet.")
-    if input.is_sparse:
-        return UsageError("Comparison for sparse tensors is not supported yet.")
+
+    if input.layout not in {torch.strided, torch.sparse_coo, torch.sparse_csr}:
+        return UsageError(f"Unsupported tensor layout {input.layout}")
```

Comment:
This currently leaves 

```python
print(
    [
        name
        for name, value in torch.__dict__.items()
        if isinstance(value, torch.layout)
        and value not in {torch.strided, torch.sparse_coo, torch.sparse_csr}
    ]
)
```

```
['_mkldnn']
```

as only unsupported `layout`. The documentation does not say a lot about MKLDNN. Does it need special handling or is it compatible with `torch.strided`?

Reply:
> Does it need special handling or is it compatible with `torch.strided`?

It is not compatible with `torch.strided`. @mruberry do you think that it is worth it to add support for it given the layout is "private" (prefixed with an underscore) and the documentation is rather scarce?

Reply:
Definitely not in this PR -- probably not later unless we get an explicit request to add it, either